pkg/search/postgres/store_cache.go

```diff
@@ -0,0 +1,386 @@
+		underlyingStore: underlyingStore,
+
+		setCacheOperationDurationTime: setCacheOperationDurationTime,
+	}
+	// Initial population of the cache. Make sure it is in sync with the DB.
+	store.repopulateCache()
+	return store
+}
+
+// NewGenericStoreWithCacheAndPermissionChecker returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericStoreWithCacheAndPermissionChecker[T any, PT clonedUnmarshaler[T]](
+	db postgres.DB,
+	schema *walker.Schema,
+	pkGetter primaryKeyGetter[T, PT],
+	insertInto inserter[T, PT],
+	copyFromObj copier[T, PT],
+	setAcquireDBConnDuration durationTimeSetter,
+	setPostgresOperationDurationTime durationTimeSetter,
+	setCacheOperationDurationTime durationTimeSetter,
+	checker walker.PermissionChecker,
+) Store[T, PT] {
+	underlyingStore := NewGenericStoreWithPermissionChecker[T, PT](
+		db,
+		schema,
+		pkGetter,
+		insertInto,
+		copyFromObj,
+		setAcquireDBConnDuration,
+		setPostgresOperationDurationTime,
+		checker,
+	)
+	store := &cachedStore[T, PT]{
+		schema:            schema,
+		pkGetter:          pkGetter,
+		permissionChecker: checker,
+		cache:             make(map[string]PT),
+		underlyingStore:   underlyingStore,
+
+		setCacheOperationDurationTime: setCacheOperationDurationTime,
+	}
+	// Initial population of the cache. Make sure it is in sync with the DB.
+	store.repopulateCache()
+	return store
+}
+
+// cachedStore implements subset of Store interface for resources with single ID.
+type cachedStore[T any, PT clonedUnmarshaler[T]] struct {
+	schema                        *walker.Schema
+	pkGetter                      primaryKeyGetter[T, PT]
+	setCacheOperationDurationTime durationTimeSetter
+	permissionChecker             walker.PermissionChecker
+	targetResource                permissions.ResourceMetadata
+	underlyingStore               Store[T, PT]
+	cache                         map[string]PT
+	cacheLock                     sync.RWMutex
+}
+
+// Upsert saves the current state of an object in storage.
+func (c *cachedStore[T, PT]) Upsert(ctx context.Context, obj PT) error {
+	dbErr := c.underlyingStore.Upsert(ctx, obj)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.Upsert)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	c.addToCacheNoLock(obj.Clone())
+	return nil
+}
+
+// UpsertMany saves the state of multiple objects in the storage.
+func (c *cachedStore[T, PT]) UpsertMany(ctx context.Context, objs []PT) error {
+	dbErr := c.underlyingStore.UpsertMany(ctx, objs)
+	if dbErr != nil {
+		return dbErr
+	}
+	defer c.setCacheOperationDurationTime(time.Now(), ops.UpdateMany)
+	c.cacheLock.Lock()
+	defer c.cacheLock.Unlock()
+	for _, obj := range objs {
+		c.addToCacheNoLock(obj.Clone())
+	}
+	return nil
+}
+
+// Delete removes the object associated to the specified ID from the store.
+func (c *cachedStore[T, PT]) Delete(ctx context.Context, id string) error {
+	obj, found, err := c.Get(ctx, id)
```

Comment:
This is more expensive than it needs to be as it will clone the the obj. Can we just check existence?

Reply:
Ah, I see we use it below to check write allowed

Reply:
Also, do we check that writes are scoped anywhere? I see in the deployment datastore that write access is global and not scoped to objects

Reply:
we do not need to clone the object, though, and don't need to expose to IDs being filtered out because there is no read permission on a given object while there is write permission to it (most likely pure theory, but let's assume it is possible).
Changing this to direct map access.